File: benchmarks/binary_trees/binary_trees.go
// binary_trees — allocation-churn benchmark (mirrors binary_trees.c / .rl).
// maxDepth=14; nodes are heap-allocated pointers reclaimed by Go's GC.
package main

import "fmt"

type Tree struct {
	left, right *Tree
}

func makeTree(depth int) *Tree {
	if depth == 0 {
		return &Tree{}
	}
	return &Tree{left: makeTree(depth - 1), right: makeTree(depth - 1)}
}

func check(t *Tree) int64 {
	if t == nil {
		return 0
	}
	return 1 + check(t.left) + check(t.right)
}

func main() {
	minDepth := 4
	maxDepth := 14
	var total int64 = 0

	// stretch tree
	stretchDepth := maxDepth + 1
	stretch := makeTree(stretchDepth)
	total += check(stretch)

	// long-lived tree
	longLived := makeTree(maxDepth)

	// iteration loop
	for depth := minDepth; depth <= maxDepth; depth += 2 {
		exp := maxDepth - depth + minDepth
		iterations := int64(1) << uint(exp)
		for i := int64(0); i < iterations; i++ {
			t := makeTree(depth)
			total += check(t)
		}
	}

	// long-lived tree check
	total += check(longLived)

	fmt.Println(total)
}

File: benchmarks/binary_trees/binary_trees.js
// binary_trees — allocation-churn benchmark (mirrors binary_trees.c / .rl).
// maxDepth=14; nodes are heap objects reclaimed by V8's GC.
"use strict";

function make(depth) {
  if (depth === 0) return { left: null, right: null };
  return { left: make(depth - 1), right: make(depth - 1) };
}

function check(t) {
  if (t === null) return 0;
  return 1 + check(t.left) + check(t.right);
}

function main() {
  const minDepth = 4;
  const maxDepth = 14;
  let total = 0;

  // stretch tree
  const stretchDepth = maxDepth + 1;
  const stretch = make(stretchDepth);
  total += check(stretch);

  // long-lived tree
  const longLived = make(maxDepth);

  // iteration loop
  for (let depth = minDepth; depth <= maxDepth; depth += 2) {
    const exp = maxDepth - depth + minDepth;
    const iterations = 1 << exp;
    for (let i = 0; i < iterations; i++) {
      const t = make(depth);
      total += check(t);
    }
  }

  // long-lived tree check
  total += check(longLived);

  console.log(total);
}

main();
